support/agent/miner_agent.go

```diff
@@ -0,0 +1,338 @@
+		From:          ma.Worker,
+		To:            ma.RobustAddress,
+		Value:         big.Zero(),
+		Method:        builtin.MethodsMiner.ProveCommitSector,
+		Params:        &params,
+		ReturnHandler: ma.handleProveCommit,
+	}
+}
+
+// register an op for next epoch (after batch prove) to schedule a post for the sector
+func (ma *MinerAgent) handleProveCommit(v *vm.VM, msg Message, _ cbor.Marshaler) error {
+	sn := msg.Params.(*miner.ProveCommitSectorParams).SectorNumber
+	ma.PreCommits--
+	ma.operationSchedule.ScheduleOp(v.GetEpoch(), registerSectorAction{sectorNumber: sn})
+	return nil
+}
+
+// create precommit message and activation trigger
+func (ma *MinerAgent) createPrecommit(currentEpoch abi.ChainEpoch, sectorNumber abi.SectorNumber, sectorActivation abi.ChainEpoch) Message {
+	params := miner.PreCommitSectorParams{
+		SealProof:     ma.Config.ProofType,
+		SectorNumber:  sectorNumber,
+		SealedCID:     sectorSealCID(ma.rnd),
+		SealRandEpoch: currentEpoch - 1,
+		Expiration:    ma.sectorExpiration(currentEpoch),
+	}
+	return Message{
+		From:   ma.Worker,
+		To:     ma.RobustAddress,
+		Value:  big.Zero(),
+		Method: builtin.MethodsMiner.PreCommitSector,
+		Params: &params,
+		ReturnHandler: func(_ *vm.VM, _ Message, _ cbor.Marshaler) error {
+			ma.PreCommits++
+			ma.operationSchedule.ScheduleOp(sectorActivation, proveCommitAction{sectorNumber})
+			return nil
+		},
+	}
+}
+
+// create a random valid sector expiration
+func (ma *MinerAgent) sectorExpiration(currentEpoch abi.ChainEpoch) abi.ChainEpoch {
+	// these are precommit bounds. Prove commit is more lenient but contains this range.
+	minExp := currentEpoch + miner.MaxProveCommitDuration[ma.Config.ProofType] + miner.MinSectorExpiration
+	maxExp := currentEpoch + miner.MaxSectorExpirationExtension
+	return minExp + abi.ChainEpoch(ma.rnd.Int63n(int64(maxExp-minExp)))
+}
+
+// Generate a sector activation over the range of acceptable values.
+// The range varies widely from 150 - 3030 epochs after precommit.
+// Assume differences in hardware and contention in the miner's sealing queue create a uniform distribution
+// over the acceptable range
+func (ma *MinerAgent) sectorActivation(preCommitAt abi.ChainEpoch) abi.ChainEpoch {
+	minActivation := preCommitAt + miner.PreCommitChallengeDelay + 1
+	maxActivation := preCommitAt + miner.MaxProveCommitDuration[ma.Config.ProofType]
+	return minActivation + abi.ChainEpoch(ma.rnd.Int63n(int64(maxActivation-minActivation)))
+}
+
+// compute next precommit according to a poisson distribution
+func precommitDelay(rate float64, rnd *rand.Rand) float64 {
+	return -math.Log(1-rnd.Float64()) / rate
+}
+
+// create a random seal CID
+func sectorSealCID(rnd *rand.Rand) cid.Cid {
+	data := make([]byte, 10)
+	_, err := rnd.Read(data)
+	if err != nil {
+		panic(err)
+	}
+
+	sum := sha256.Sum256(data)
+	hash, err := mh.Encode(sum[:], miner.SealedCIDPrefix.MhType)
+	if err != nil {
+		panic(err)
+	}
+	return cid.NewCidV1(miner.SealedCIDPrefix.Codec, hash)
```

Comment:
Do we need these CIDs to be this "real"? Could we just use a Raw/Identity CID instead?

In any case I think this should use a `cid.V1Builder` rather than do the SHA and encoding manually. See example in `builtin/codes.go`

Reply:
We actually do. At least we need them to have the right Cid prefix because we validate it in PreCommit. I would like to use the `miner.SealedCIDPrefix.Sum` function, but it panics because it's not-implemented. This is the simplest solution I've found so far that works.